### ISSUE TITLE
Make it possible to store a preferred set of hooks as R options?

### ISSUE THREAD

**Issue by pat-s** [lorenzwalthert/precommit#85]
With more hooks becoming available, one might often want to use the same set of hooks for multiple repos.

It would be convenient to store a default set of hooks as R options which will be enabled in a repo if one calls `precommit::use_precommit()` in a new repo.

#55 sounds related. Just close if you consider this a duplicate.

**Comment by lorenzwalthert** [comment]
#55 is different I think, this would simply make `pre-commit install` obsolete, without touching any config file. 

Maybe we could introduce an R option that stores a path to a template config file or allow to specify a configuration we can `yaml::write_yaml()` from a list and store this list in an R option. Drawback: Cannot contain comments (`#`) as we use them now.

**Comment by pat-s** [comment]
This is tricky, yes.

How about an option/env var pointing to a pre-commit.yaml stored on Github (raw) which will be used as the template for installing and updating? If not set, it will use the template from {precommit}.

In addition to the pointer, maybe an option specifying if `precommit::autoupdate()` should also use the template for updating eventual missing hooks.

Why this?
When changing a hook in the template set, I usually want this hook to be activated in all other repos as well when running `precommit::autoupdate()` rathern than having to uninstall/install again.

```r
precommit.template = <link>
precommit.autoupdate.use.template = TRUE/FALSE
```

**Comment by lorenzwalthert** [comment]
I think this feature request may actually fit better in the upstream repo. Because it's not R specific. The CLI could be extended like this:

-  `pre-commit install --config https://raw.githubusercontent.com/path/to/raw/.pre-commit-config.yaml`
- `pre-commit install --config file:///abs/path/to/config-1.yaml`.

The file name should be specified, then different files could be stored, and people could use different config files that are stored in the same repo.

Then, `use_precommit()` could expose `config_dir` and `config_file`, both defaulting to respective R options. The latter, if missing, should fall back to `.pre-commit-config.yaml`. What do you think? 

The autoupdate problem should be addressed separately I think. I think the clean way would be to store the template reference in the config file, and then adapt the upstream cli again, e.g. `pre-commit autoupdate` should by default also check if there is a template config linked in the `.pre-commit-config.yaml`. If yes, it should update, unless `--no-template-sync` or similar is given.

**Comment by pat-s** [comment]
> Then, use_precommit() could expose config_dir and config_file, both defaulting to respective R options. The latter, if missing, should fall back to .pre-commit-config.yaml. What do you think?

Sounds good. 

> The CLI could be extended like this:

Meaning currently this is not possible yet and needs to be added upstream first?

> The autoupdate problem should be addressed separately I think. 

Agree.

**Comment by lorenzwalthert** [comment]
Yes, in my understanding, pre-commit expects you th have a config file in the root at the time you `pre-commit install`.

**Comment by lorenzwalthert** [comment]
After the in-depth discussion in https://github.com/pre-commit/pre-commit/issues/1308, I don't think this needs further consideration. I prefer to keep this repo lightweight and I don't think we should add significant, non-R related features here, at least unless there are ways to solve these problems with the upstream framework.

**Comment by pat-s** [comment]
What is the best way to achieve this with {precommit} right now?

How about exposing a `config` arg to `use_precommit()`?
This could check for a default location (e.g. `~/.config/pre-commit/config.yaml`) and use it if it exists (via the `--config` flag of `pre-commit`)?

**Comment by lorenzwalthert** [comment]
Like asottile, I don't like these approaches so much. In particular the way the `--config` approach is implemented is intransparent to me because I'd prefer to have one config file per repo and track it in git in this very repo, not a centralised one someone once linked from his local machine. 

My hesitations with using the `config` argument in `use_precommit()` and simply transfer a config from elsewhere is that people who know pre-commit will think it's equivalent to `pre-commit install --config`, which is where the trouble starts, because it's not. It's a API design principle to make it hard for people to get surprised, and such a feature would violate this principle in my opinion. 

I think the best way to achieve what you want for a large number of repos once is similar to this (not tested!):

```r
use_custom_precommit_setup <- function(roots, source = "~/.pre-commit-config.yaml") {
  fs::file_copy(source, roots)
  purrr::walk(roots, ~withr::with_dir(.x, {
    precommit::use_precommit(force = FALSE)
  })
}
```

Using a different name for this functionality than `use_precommit` mitigates the danger zone as far as API design goes. We might consider adding such a function to `{precommit}`. What do you think? Sorry I am changing my mind all the time haha.

**Comment by lorenzwalthert** [comment]
Closed with #111.
